Rewrite the local changesets in-place for client reset recovery

tgoyne opened the pull request:
Rather than discarding the old sync history and creating new commits, update the individual changesets which are being recovered in-place while keeping the original version numbers. This significantly simplifies recovery when pending subscriptions are involved, as they no longer need to be updated at all, and it lets us apply the client reset in a single write transaction, which improves notifications and avoids a lot of problems if another thread has the Realm open while the client reset is happening.

It also avoids causing problems for the server, which doesn't like overly large changesets, and sometimes ran into problems due to all of the pending local changesets being merged into one.

---------------------------

"Move RecoverLocalChangesetsHandler fully to the cpp file" is just moving code around with no changes, and the functional changes to client_reset_recovery.cpp are entirely in the final commit.

"Remove SubscriptionStore::get_mutable_by_version()" is adjusting the SubscriptionStore API to eliminate a problem I ran into while writing some tests: the MutableSubscriptionSet it returns doesn't actually support modifying the subscription set. This is actually the correct behavior, as existing subscription sets shouldn't be modified, but it was a misleading interface. I changed it to only expose `update_state()` rather than appearing to allow more than it did.

The client reset during async open tests started hitting some very unintended codepaths and started crashing. AFAICT the change in timing resulted in it triggering two client resets, as it relied on stopping sync in the middle of the server sending a batch having some very specific behavior that wasn't actually guaranteed. I updated it to use the client reset endpoint that all other tests use and to immediately disconnect rather than relying on the server disconnecting it. This revealed that the FLX test had failed to test the thing it was trying to test and that very specific scenario was broken.

ironage reviewed (approved):
Nice improvements across the board 👍 
It is a lot of code changes in some dusty corners of the codebase, but I went through all the commits and it looks good to me. Thanks for taking this on, having the FLX recovery as a single commit is a huge win!

danieltabacaru commented on src/realm/sync/subscriptions.cpp:
```diff
@@ -859,18 +817,59 @@ void SubscriptionStore::terminate()
+    switch (new_state) {
+        case State::Error:
+            if (old_state == State::Complete) {
+                throw RuntimeError(
```
I think this can happen if a queryable field is removed after the query is marked complete

tgoyne replied:
I'm not sure what we should do when that happens. The SubscriptionStore code assumes that the active subscription set is always valid and there isn't an obvious resolution when a query which used to be valid no longer is. This change is just adjusting the error message and code which is reported when it does happen.

tgoyne replied:
I'll add a note about this possibly due to a queryable field being removed to the error.

danieltabacaru commented on src/realm/sync/subscriptions.cpp:
```diff
@@ -859,18 +817,59 @@ void SubscriptionStore::terminate()
+void SubscriptionStore::update_state(int64_t version, State new_state, std::optional<std::string_view> error_str)
 {
+    REALM_ASSERT(error_str.has_value() == (new_state == State::Error));
+    REALM_ASSERT(new_state != State::Uncommitted);
```
you already cover the assertions on new_state in the switch statement

danieltabacaru commented on src/realm/sync/noinst/client_history_impl.cpp:
```diff
@@ -37,58 +38,88 @@
-    prepare_for_write();             // Throws
+        util::compression::CompressMemoryArena arena;
+        util::AppendBuffer<char> compressed;
+        for (auto& [changeset, version] : recovered_changesets) {
```
I think for each recovered changeset we need to purge its reciprocal transform (if it exists) since they will be part of the merge window.

tgoyne replied:
Good catch, there definitely needs to be something done with reciprocal transforms. I'm not sure discarding them is correct. I may just be misunderstanding what reciprocal transforms are, but would it make sense for `ClientHistory::get_local_changes()` to return them instead of the original changeset when they're present?

danieltabacaru replied:
You can have a read on reciprocal history [here](https://github.com/10gen/baas/blob/master/realm/docs/3_ot.md#16-reciprocal-history). I am not sure using the reciprocal changes is correct either (at the time a client reset happens there may still be server changes that could have updated them, but once the client reset is done those changes are integrated without having had any impact on the reciprocal changes). This may be the reason why this approach won't work (although I am not fully convinced yet this was not an issue before too). I will give it more thought.

tgoyne replied:
After poking at the actual results it gives I'm pretty confident that recovering the reciprocal history is better than recovering the original changesets. Obviously this is all a best-effort thing that can never be fully correct, but using the reciprocal history puts the rebase origin point at the last integrated DOWNLOAD message rather than the DOWNLOAD message received before the write was originally made.

danieltabacaru replied:
indeed, that's what I was trying to explain above. it would be nice to add those tests.

tgoyne commented on src/realm/sync/noinst/client_history_impl.cpp:
```diff
@@ -37,58 +38,88 @@
+        do_trim_sync_history(discard_count);
+
+        if (logger.would_log(util::Logger::Level::debug)) {
+            logger.debug("Client reset adjustments: trimming %1 history entries and updating %2 of %3 remaining "
```
There's often going to be some history entries with empty changesets that aren't getting updated.

danieltabacaru replied:
right

tgoyne commented on test/test_sync_subscriptions.cpp:
```diff
@@ -165,21 +165,18 @@ TEST(Sync_SubscriptionStoreStateUpdates)
-        // By marking version 2 as complete version 1 will get superceded and removed.
-        CHECK_THROW(store->get_mutable_by_version(1), KeyNotFound);
+        // By marking version 2 as complete version 1 will get superseded and removed.
+        CHECK_EQUAL(store->get_by_version(1).state(), SubscriptionSet::State::Superseded);
```
"superseded and removed" is a redundant phrase, as we mark things superseded by removing them. The comment is probably misleading, but the difference in the test is just due to `get_by_version()` and `get_mutable_by_version()` reporting a superseded version in different ways.

tgoyne commented:
I've switched it over to recovering using the reciprocal history, made it clear the reciprocal history afterwards, and added a test showing why each of those is needed.

danieltabacaru commented on src/realm/sync/subscriptions.cpp:
```diff
@@ -973,4 +971,25 @@ int64_t SubscriptionStore::set_active_as_latest(Transaction& wt)
     return version;
 }
 
+int64_t SubscriptionStore::mark_active_as_complete(Transaction& wt)
```
you could make this a no-op if the state is already Complete (unless it's possible to set the state and not deliver the notifications)

tgoyne replied:
I think if the active set is already complete then `m_pending_notifications` should be empty barring someone waiting for a Superseded notification, so the only unnecessary thing we're doing in that scenario is acquiring the lock.

danieltabacaru commented on test/object-store/util/sync/sync_test_utils.cpp:
```diff
@@ -507,39 +507,16 @@ void wait_for_num_objects_in_atlas(std::shared_ptr<SyncUser> user, const AppSess
 }
 
-void trigger_client_reset(const AppSession& app_session)
+void trigger_client_reset(const AppSession& app_session, const SyncSession& sync_session)
```
don't we need to set development mode back and wait for initial sync to complete as before? (is the client reset endpoint doing all that?)

tgoyne replied:
The client reset endpoint just directly triggers a client reset for a specific client file ident without disabling sync, so none of that is needed.

danieltabacaru reviewed (approved):
LGTM. You could mention using the reciprocal history in the PR's description.
